runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -1190,10 +1190,42 @@ class INTERPRETER_CLASS
 					_currentThread->tempSlot = (UDATA)invalidIndex;
 					rc = THROW_AIOB;
 				} else {
-					I_32 value = VM_ArrayCopyHelpers::referenceArrayCopy(_currentThread, srcObject, srcStart, destObject, destStart, elementCount);
-					if (-1 != value) {
-						buildInternalNativeStackFrame(REGISTER_ARGS);
-						rc = THROW_ARRAY_STORE;
+#if defined(J9VM_OPT_VALHALLA_VALUE_TYPES)
+					J9Class *srcClazz = J9OBJECT_CLAZZ(_currentThread, srcObject);
+					J9Class *destClazz = J9OBJECT_CLAZZ(_currentThread, destObject);
+					J9Class *destComponentClass = ((J9ArrayClass *)destClazz)->componentType;
+
+					if (J9_IS_J9CLASS_FLATTENED(srcClazz) && J9_IS_J9CLASS_FLATTENED(destClazz)) {
```

Comment:
I think this has to be, if both are flattened and dont contain any object references. If there is an object ref, you need to do a barrier load/store

Reply:
We should consider making a `VM_ArrayCopyHelpers::vtArrayCopy` that does a batch barrier. @dmitripivkine

Reply:
You can check if the class has references between doing `J9_ARE_ANY_BITS_SET(clazz->classFlags, J9ClassHasReferences)`

Reply:
Ok I updated the check in 9f31ca4